_posts/2019-11-04-generic-programming-part-2-implementation-overview.markdown

```diff
@@ -0,0 +1,603 @@
+            return second;
+        }
+        return first;
+    }
+}
+```
+
+
+#### Under the hood {#cpp_templates_under_the_hood}
+
+
+Word **Template** reveals how does language feature work under the hood.
+Implemented function is just a template for compiler,
+which generates new functions for every type template is used with.
+
+So If you use `max` function like this:
+```cpp
+cout << "int " << generic::max(1, 2) << ", char " << generic::max('a', 'b');
+```
+under the hood compiler generates 2 functions:
+
+```cpp
+template<> int max<int>(int first, int second) {
+    if (second > first) {
+        return second;
+    }
+    return first;
+}
+template<> char max<char>(char first, char second) {
+    if (second > first) {
+        return second;
+    }
+    return first;
+}
+```
+
+#### Migration to generics {#cpp_migration_to_generics}
+
+I don't believe that migration have ever been an issue in C++.
+Templates doesn't break existing code.
+And in C++ even now*(year 2019)* many projects are written without usage of any library.
+For some reason C++ developers like to implement everything by themselves.
+Anyway first release of STL(Standard Template Library) where generic code really necessary was at 1992,
+and library code based on templates.
+
+#### Advantages {#cpp_templates_advantages}
+
+Approach from C++ is the most understandable for developers,
+it's easy to imagine how code would work if you replace generic parameter with the specific type.
+It's really easy work with generic types - you can do what ever you want with them,
+all type checks take place after code generation phase,
+just write code so that it compile with specific type you use it with.
+
+Templates are also very flexible,
+you can use just anything inside template:
+it would work if it works with template parameter.
+
+#### Disadvantages {#cpp_templates_disadvantages}
+
+But templates aren't very popular in C++ world.
+Many developers don't use that feature.
+There is too many disadvantages:
+
+* Templates slows down compilation;
+* Absolutely not helpful compiler error messages;
+* You can't reuse generic binary, only source code; 
+* Hard to validate: to make sure it works with type you need to test it with that type;
+* Code bloat.
+
+**Code bloat** it's when you get much bigger executable file after compilation then you expected.
+Every time you use a new generic parameter you get a new generated class or function,
+its code is stored on disk and have to be loaded in the RAM during execution.
+
+#### Variance {#cpp_templates_variance}
+As I sad before C++ templates are very intuitive:
+there is just code generation behind it.
+What kind of variance will it be if you write the same class or function using different types?
+**Invariance**.
+There is no relationships between generated classes and functions.
+
+But since C++ 17 std library supports co and contravariance via `std::function`.
+Functions produces are covariant, functions consumers are contravariant.
```

Comment:
Typo and producers are covariative( we need an adjective), consumers are contravariative

Reply:
As I see in [Wikipedia](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)) people use **covariant** as adjective, and  **covariance** as noun.